docs/extending/extending_draftail.md

```diff
@@ -329,17 +325,203 @@ To fully complete the demo, we can add a bit of JavaScript to the front-end in o
+  }, `Sentences: ${countSentences(text)}`);
+}
+
+window.draftail.registerPlugin({
+  type: 'sentences',
+  meta: SentenceCounter,
+}, 'controls');
+```
+
+### Text decorators
+
+The [decorators API](https://www.draftail.org/docs/decorators) is how Draftail / Draft.js supports highlighting text with special formatting in the editor. It uses the [CompositeDecorator](https://draftjs.org/docs/advanced-topics-decorators/#compositedecorator) API, with each entry having a `strategy` function to determine what text to target, and a `component` function to render the decoration.
+
+There are two important considerations when using this API:
+
+- Order matters: only one decorator can render per character in the editor. This includes any entities that are rendered as decorations.
+- For performance reasons, Draft.js only re-renders decorators that are on the currently-focused line of text.
+
+Here is an example with highlighting of problematic punctuation – first, registering the editor feature in a `wagtail_hooks.py`:
+
+```python
+from wagtail.admin.rich_text.editors.draftail.features import DecoratorFeature
+
+
+@hooks.register('register_rich_text_features')
+def register_punctuation_highlighter(features):
+    feature_name = 'punctuation'
+    features.default_features.append(feature_name)
+
+    features.register_editor_plugin(
+        'draftail',
+        feature_name,
+        DecoratorFeature({
+            'type': feature_name,
+        },
+            js=['draftail_punctuation.js'],
+        ),
+    )
+```
+
+Then, `draftail_punctuation.js` defines the strategy and the highlighting component:
+
+```javascript
+const PUNCTUATION = /(\.\.\.|!!|\?!)/g;
+
+const punctuationStrategy = (block, callback) => {
+  const text = block.getText();
+  let matches;
+  while ((matches = PUNCTUATION.exec(text)) !== null) {
+    callback(matches.index, matches.index + matches[0].length);
+  }
+};
+
+const errorHighlight = {
+  color: 'var(--w-color-text-error)',
+  outline: '1px solid currentColor',
+}
+
+const PunctuationHighlighter = ({ children }) => (
+  window.React.createElement('span', { style: errorHighlight, title: 'refer to our styleguide' }, children)
+);
+
+window.draftail.registerPlugin({
+  type: 'punctuation',
+  strategy: punctuationStrategy,
+  component: PunctuationHighlighter,
+}, 'decorators');
+```
+
+### Arbitrary plugins
+
+```{warning}
+This is an advanced feature. Please carefully consider whether you really need this.
+```
+
+Draftail supports plugins following the [Draft.js Plugins](https://www.draft-js-plugins.com/) architecture. Such plugins are the most advanced and powerful type of extension for the editor, offering customisation capabilities equal to what would be possible with a custom Draft.js editor.
+
+A common scenario where this API can help is to add bespoke copy-paste processing. Here is a simple example, automatically converting URL anchor hash references to links. First, let’s register the extension in Python:
```

Comment:
There’s lots of options for what to demonstrate here. I chose this because copy-paste processing feels like a relatively universal improvement, and keeps the line count of the example _relatively_ small.